src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -115,16 +113,182 @@ public CodeGenIL(IDictionary<string, IList> input, string filePath)
+
             if(argType == typeof(double) && param.ParameterType == typeof(long))
             {
                 EmitOpCode(OpCodes.Conv_I8);
                 return typeof(long);
             }
+            if (argType == typeof(double) || argType == typeof(long) && param.ParameterType == typeof(int))
+            {
+                EmitOpCode(OpCodes.Conv_I4);
+                return typeof(int);
+            }
+            if(argType == typeof(double[]) && typeof(IEnumerable<int>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<double, int>();
+            }
+            if(argType == typeof(int[]) && typeof(IEnumerable<double>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<int, double>();
+            }
+            if (argType == typeof(double[]) && typeof(IEnumerable<long>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<double, long>();
+            }
+            if (argType == typeof(long[]) && typeof(IEnumerable<double>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<long, double>();
+            }
+            // TODO: The following cases do not work yet as EmitIListCoercion needs to be fixed.
+            //if (typeof(IList).IsAssignableFrom(argType) && typeof(IEnumerable<double>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<double>();
+            //}
+            //if (typeof(IList).IsAssignableFrom(argType)  && typeof(IEnumerable<int>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<int>();
+            //}
+            //if (typeof(IList).IsAssignableFrom(argType) && typeof(IEnumerable<long>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<long>();
+            //}
             // TODO: Add more coercion cases here.
 
             return argType;
         }
 
+        // TODO: Does not work yet (intended to coerce IList to IEnumerable<T>)
+        private Type EmitIListCoercion<T>()
+        {
+            MethodInfo mInfo = null;
+            // Find length for IList to be coerced (already on top of eval stack), len
+            var prop = typeof(ICollection).GetProperties(BindingFlags.Public | BindingFlags.Instance).Where(
+                                p => p.Name == nameof(ICollection.Count)).FirstOrDefault();
+
+            if (prop != null)
+            {
+                mInfo = prop.GetAccessors().FirstOrDefault();
+            }
+            //var mInfo = typeof(ICollection).GetMethod(nameof(ICollection.get_Count));
+            EmitOpCode(OpCodes.Callvirt, mInfo);
+
+            // var newarr = new T[len];
+            EmitOpCode(OpCodes.Newarr, typeof(T));
+
+            // Declare new array to store coerced values
+            var t = typeof(T).MakeArrayType();
```

Comment:
but a `system.array` does not implement `IEnumerable<T>` ?

Reply:
This basically makes type `T` into type `T[]`.

Reply:
sorry, should have been more clear - but `T[]` does not implement `IEnumerable<T>` so it's not clear how this method could work to convert `IList` to `IEnumerable<T>`

Reply:
It doesn't, I attempted writing `EmitIListCoercion` to do that, which I haven't gotten to work yet.

Reply:
Here, I'm emitting code to initialize a new array of type `T[]`, then take, the `IList` and call `CopyTo` on it to copy its elements into the new array like: `ilist.CopyTo(newarr, 0);`

I think `IEnumerable<T>` is assignable from `T[]`, which is why I chose to convert the IList to `T[]`.

Reply:
@aparajit-pratap you are correct:
```
Single-dimensional arrays implement the [System.Collections.Generic.IList<T>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ilist-1?view=net-6.0), [System.Collections.Generic.ICollection<T>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.icollection-1?view=net-6.0), [System.Collections.Generic.IEnumerable<T>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ienumerable-1?view=net-6.0), [System.Collections.Generic.IReadOnlyList<T>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlylist-1?view=net-6.0) and [System.Collections.Generic.IReadOnlyCollection<T>](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlycollection-1?view=net-6.0) generic interfaces. The implementations are provided to arrays at run time, and as a result, the generic interfaces do not appear in the declaration syntax for the [Array](https://docs.microsoft.com/en-us/dotnet/api/system.array?view=net-6.0) class. In addition, there are no reference topics for interface members that are accessible only by casting an array to the generic interface type (explicit interface implementations). The key thing to be aware of when you cast an array to one of these interfaces is that members which add, insert, or remove elements throw [NotSupportedException](https://docs.microsoft.com/en-us/dotnet/api/system.notsupportedexception?view=net-6.0).
```

it even does implement those interfaces for single dimension arrays, but it's injected at runtime... what?

Reply:
Unfortunately, however, I'm getting an invalidcastexception, which I cannot seem to understand why. When I try the same exact code in linqpad it works.